Update Published App Services - Fall 2019 Documentation (#170)
smartdevicelink/sdl_app_library_guides: docs/Other SDL Features/Creating an App Service/index.md

> # Creating an App Service

Reviewer: ```suggestion
# Creating an App Service (RPC v5.1+)
```

> App services is a powerful feature enabling both a new kind of vehicle-to-app communication and app-to-app communication via SDL. App services is available on head units supporting RPC v5.1+. 

Reviewer: ```suggestion
App services is a powerful feature enabling both a new kind of vehicle-to-app communication and app-to-app communication via SDL.
```

Using an app service is covered [in another guide](Other SDL Features/Using App Services).

## App Service Types
Apps are able to declare that they provide an app service by publishing an app service manifest. Three types of app services are currently available and more will be made available over time. The currently available types are: Media, Navigation, and Weather. An app may publish multiple services (one for each of the different service types) if desired.

## Publishing an App Service

> Publishing a service is a multistep process. First, you need to create your app service manifest. Second, you will publish your app service. Third, you will publish the service data using `OnAppServiceData`. Fourth, you must listen for requests for data and respond accordingly. Fifth, if your app service supports handling of RPCs related to your service you must listen for these requests and handle them accordingly. Sixth, optionally, you can support URI based app actions. Finally, if necessary, you can you can update or delete your app service manifest.

Reviewer: ```suggestion
Publishing a service is a multi-step process. First, you need to create your app service manifest. Second, you will publish your app service to the module. Third, you will publish the service data using `OnAppServiceData`. Fourth, you must listen for data requests and respond accordingly. Fifth, if your app service supports handling of RPCs related to your service you must listen for these RPC requests and handle them accordingly. Sixth, optionally, you can support URI-based app actions. Finally, if necessary, you can you can update or delete your app service manifest.
```

});
```
!@

## Updating Your Published App Service

> Once you have published your app service, you may decide to update the type of data provided in your app service. For example, you may want to give all-access to paid subscribers but provide limited information to free users. You can do this by updating the app service manifest. If desired, you can also delete your app service by unpublishing the service. 

Reviewer: ```suggestion
Once you have published your app service, you may decide to update its data. For example, if you have a free and paid tier with different amounts of data, you may need to upgrade or downgrade a user between these tiers and provide new data in your app service manifest. If desired, you can also delete your app service by unpublishing the service. 
```

### 7. Updating a Published App Service Manifest (RPC v6.0+)
@![iOS]
##### Objective-C
```objc
SDLAppServiceManifest *manifest = [[SDLAppServiceManifest alloc] initWithAppServiceType:SDLAppServiceTypeWeather];
manifest.weatherServiceManifest = <#Updated weather service manifest#>

SDLPublishAppService *publishServiceRequest = [[SDLPublishAppService alloc] initWithAppServiceManifest:manifest];
[self.sdlManager sendRequest:publishServiceRequest];
```

##### Swift
```swift
let manifest = SDLAppServiceManifest(appServiceType: .weather)
manifest.weatherServiceManifest = <#Updated weather service manifest#>

let publishServiceRequest = SDLPublishAppService(appServiceManifest: manifest)
sdlManager.send(publishServiceRequest)
```
!@

@![android,javaSE,javaEE]

> ##### Java

Reviewer: ```suggestion
```

They don't usually provide this since they only have one language, though I see it in this document so... ¯\_(ツ)_/¯

Probably should remove it throughout the document.

```java
AppServiceManifest manifest = new AppServiceManifest(AppServiceType.WEATHER);
manifest.setWeatherServiceManifest("Updated weather service manifest");

PublishAppService publishServiceRequest = new PublishAppService(manifest);
sdlManager.sendRPC(publishServiceRequest);
```
!@

### 8. Unpublishing a Published App Service Manifest (RPC v6.0+)
@![iOS]
##### Objective-C
```objc
SDLUnpublishAppService *unpublishAppService = [[SDLUnpublishAppService alloc] initWithServiceID:@"<#The serviceID of the service to unpublish#>"];
[self.sdlManager sendRequest:unpublishAppService];
```

##### Swift
```swift
let unpublishAppService = SDLUnpublishAppService(serviceID: "<#The serviceID of the service to unpublish#>")
sdlManager.send(unpublishAppService)
```
!@

@![android,javaSE,javaEE]

> ##### Java

Reviewer: See above

Author: Fixed. Removed all Java headers for the code snippets
